state.go

```diff
@@ -129,6 +142,24 @@ type Proposal struct {
 
 	// Time is the time to create the proposal
 	Time time.Time
+
+	// Hash is the digest of the data to seal
+	Hash []byte
```

Comment:
Should we make `Hash` a read-only field, disallowing setting it from the outside, since it is derived type of data? 
I was thinking of having `NewProposal` function which would instantiate `Proposal` and set `Hash` field based on hashing `Data` field.

Also should we validate whether `Hash` is matching `hash(Data)` or we shouldn't mind about it (although equality is checked against Hash field, so that could be the same thing implicitly at the end)?

Reply:
A couple of interesting points here. We cannot just do `hash=sha(Proposal)` because the hash has to exclude some fields of the proposal (this is because this hash is used to seal and validate the block but some fields of the block are generated after the block is sealed in pbft).

In order to solve that there was the `Hash` function in the backend that would take a proposal and hash it.

This PR does not change the fact that the backend has to be the one hashing the proposal, it only changes the place where it is done such that we can do a more meaningful validation. It is not the same to discard a block because the hash is wrong than let the pbft protocol fail silently because the Hash callback returned an incorrect Hash.